src/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs

```diff
@@ -0,0 +1,327 @@
+
+                if (counts._signalCount != 0)
+                {
+                    newCounts._signalCount--;
+                }
+                else if (timeoutMs != 0)
+                {
+                    if (_spinCount > 0 && newCounts._spinnerCount < byte.MaxValue)
+                    {
+                        newCounts._spinnerCount++;
+                    }
+                    else
+                    {
+                        // Maximum number of spinners reached, register as a waiter instead
+                        newCounts._waiterCount++;
+                        Debug.Assert(newCounts._waiterCount != 0); // overflow check, this many waiters is currently not supported
+                    }
+                }
+
+                Counts countsBeforeUpdate = Counts.CompareExchange(ref _separated._counts, newCounts, counts);
+                if (countsBeforeUpdate == counts)
+                {
+                    if (counts._signalCount != 0)
+                    {
+                        return true;
+                    }
+                    if (newCounts._waiterCount != counts._waiterCount)
+                    {
+                        return WaitForSignal(timeoutMs);
+                    }
+                    if (timeoutMs == 0)
+                    {
+                        return false;
+                    }
+                    break;
+                }
+
+                counts = countsBeforeUpdate;
+            }
+
+            int spinIndex = s_processorCount > 1 ? 0 : SpinSleep0Threshold;
+            while (spinIndex < _spinCount)
+            {
+                // TODO: Unify this with LowLevelSpinWaiter
+                if (s_processorCount > 1)
+                {
+                    if (spinIndex < SpinSleep0Threshold || (spinIndex - SpinSleep0Threshold) % 2 != 0)
+                    {
+                        // Matches algorithm in SpinWait.SpinOnce
+                        int n = RuntimeThread.OptimalMaxSpinWaitsPerSpinIteration;
+                        if (spinIndex <= 30 && (1 << spinIndex) < n)
+                        {
+                            n = 1 << spinIndex;
+                        }
+                        RuntimeThread.SpinWait(n);
+                    }
+                    else
+                    {
+                        RuntimeThread.Yield();
+                    }
+
+                    spinIndex++;
+                }
+                else
+                {
+                    // On uniprocessor system we only Yield because there cannot be other
+                    // threads executing in parallel and satisfying the condition. We also
+                    // compensate for the unnecessary spins by incrementing spinIndex by 2
+                    // and doing the same number of Yields as in the multi-processor code
+                    // above.
+                    RuntimeThread.Yield();
+                    spinIndex += 2;
+                }
+
```

Comment:
Please copy the comment from the original code: `// Try to acquire the semaphore and unregister as a spinner`